### PR TITLE
Rust performance improvements

### PR DESCRIPTION
**Pull request by kazimuth**
Hi! I was just messing around with the code and ended up doing some optimization for fun. Merge this if you want, no pressure :)

Changes:
- added benchmarks based on [criterion.rs](https://github.com/bheisler/criterion.rs); to run them use `cargo bench`
- switched random number generation to use the rand_isaac crate.
  - The rng is now stored in the `Universe` struct, and you can do `api.rand_int()` and so on. This lets you run the benchmarks; otherwise they crash because `js_sys::Math::random()` doesn't work when compiling natively. In theory it should be faster as well, but I can't measure that (because the benchmarks don't work without it).
- `#[inline]`d some functions
- changed iteration order in `tick()` from x,y to y,x so that memory accesses are sequential

According to my benchmarks this adds up to about a 55% performance improvement in `tick()`, although my benchmark is kinda dumb. This results in slightly smoother framerates when running complicated simulations. In cases where the app is gpu-limited it won't really help, though. (Maybe another optimization to investigate would be rendering at lower resolutions, but I don't know how to do that.)

### CODE REVIEW

**Comment by MaxBittker** [comment]
This is awesome @kazimuth! 

First off, having benchmarks at all is super great, and I appreciate that. It's cool to see universe being used programatically, and makes me want to write some deterministic tests for different behaviors (which would be nice to have while performance tuning!) 

Second, I haven't run this locally yet, but I expect the iteration order change, on its own, results in significantly changes behavior, because of the way that the update methods are sensitive to the order they're evaluated in (relative to their neighbors). For instance, does sand still spread vertically out as it falls, or does it drop clumped up with no air between it? I also think this would make gas trend up instead of left (I've been meaning to reverse x iteration order on consecutive ticks to address this) 
If this change was an important performance improvement, then we can change the way x and y are stored in memory to preserve the speedup!

Speaking of that "If", I think that I want to take a conservative approach to optimization here, and only merge changes that don't hurt readability, OR are proven meaningful wins. to that end, I want to answer two questions:

1)  What is the relationship between this benchmark and the performance of the game running in-browser?
(I relied on chrome and firefox's performance tracing tools during development, and that does give us a powerful if not easily automated way to measure tick execution time)  
screenshot from 2018-12-18 00-17-27

If we show once or twice that this native benchmark is a decent proxy for the performance of the wasm code, then I'll be more comfortable to rely on it for future changes.

2) Of the changes in this PR, which ones contributed to the perf wins and how much? For instance, if the unsafe block isn't significant, I definitely don't want to introduce it. It would be nice to evaluate each logical change on its own two feet both for our education and in order to avoid unnecessary complexity. 


Whew! So, I mentioned a bunch of things here. No pressure to answer these questions for me, I do appreciate the improvements you identified here on their own legs! The thing I would ask to be able to merge this change would be to simplify it to be the minimal PR needed to run the benchmark, so probably your first commit only.

**Comment by kazimuth** [comment]
Honestly I'm not completely sold on the relationship between the two as performance proxies, I just wasn't sure how else to benchmark things. I *think* I'm getting a roughly ~10fps improvement in some cases (50 -> 60fps on a middle-range laptop), but that's very unscientific. In other cases the simulation is clearly rendering-limited (I get FPS improvements if I resize the window), so 🤷‍♀️ 

A better strategy for measuring performance would probably be measuring FPS of a set of test universes in-browser. Like you say, that's not easily automated, though :/

The most improvement (assuming the benchmarks are at least somewhat informative) came from removing the two `%` operations in `set`, which I believe were unnecessary anyway since "set" is bounds-checked. That was a ~50% performance improvement.

The `unsafe` is only about a 3% improvement which isn't worth it, I agree.

I didn't think of the effects iteration order would have on the simulation at all, lol, thanks for the catch. Have you considered double-buffering to make iteration order less relevant to the simulation? It would probably change a lot of behaviors that would need re-tuning, though.

Changing iteration order is a 25% speedup; unfortunately, if I change it back and swap the memory layout, the speedup doesn't carry over. I believe this is because the wind y-axis is reversed, meaning whatever order we iterate in, either the cells or the wind will still be out-of-cache. If you rearranged the wind and cell arrays to both be stored column-major, you should be able to get this benefit.

**Comment by MaxBittker** [comment]
What I do is bump the universe size to 500x500 when testing performance to push the system. 

this PR looks well scoped now! gonna take a closer look later before merging.

It's weird modulus is so expensive! good find there. 

regarding iteration order, I think it's important to the properties of the algorithm to be able to resolve "race" type data conditions to not be double buffered (cells should agree on the state of the world while acting)
The memory layout thing is very interesting but it would be some work to rotate wind and cell data everywhere. It's cool to know the options is there though if one of us wants to tackle that in the future

**Comment by MaxBittker** [comment]
BTW, I think that graphics performance can be improved to be less screen size dependent, but I was having issues with blurry pixels on firefox. Will revisit eventually

File: crate/src/lib.rs [hunk 1]
```diff
@@ -107,6 +112,17 @@ impl<'a> SandApi<'a> {
 
         self.universe.burns[idx] = v;
     }
+    // generate a random number r such that 0 <= r < n
+    pub fn rand_int(&mut self, n: i32) -> i32 {
+        self.universe.rng.gen_range(0, n)
+    }
+    // generate a float r such that 0 <= r < 1
+    pub fn rand_float(&mut self) -> f32 {
```
Comment (MaxBittker) [inline crate/src/lib.rs]: I'll probably end up removing this and using rand_int everywhere

File: crate/src/lib.rs [hunk 2]
```diff
@@ -216,14 +232,17 @@ impl Universe {
     }
 
     pub fn new(width: i32, height: i32) -> Universe {
+        // seed from random.org
+        let mut rng: Isaac64Rng = SeedableRng::seed_from_u64(0x734f6b89de5f83cc);
```
Comment (MaxBittker) [inline crate/src/lib.rs]: the game is now deterministic - not sure if I have an opinion on that, but interesting

Comment (albinekb) [reply crate/src/lib.rs]: That could be a good thing if you want to implement snapshots like suggested here: https://www.reddit.com/r/rust/comments/a6v0p1/sandspiel_a_falling_sand_game_built_in_rustwebgl/ebyyt00/

**Comment by MaxBittker** [comment]
On a 500x500 grid, I got about 40 fps, measurably better than 30fps without the patch. Nice work!

**Comment by MaxBittker** [comment]
Hmm, it looks like depending on rand more than doubles the bundle size (weirdly seems to impact the .js as much as the .wasm).

Sorry for being such a perfectionist about this that makes pushes that aspect of the CL back into the red for me, especially because I suspect that replacing js::sys::random wasn't actually contributing to the speedup.

I once tried replacing rand_int with a hardcoded "return 0" body, and go an almost unmeasurable speedup

**Comment by MaxBittker** [comment]
I may be making a mistake measuring the bundle size, or there may be a way to get a prng without that drawback.

OR we could mock the random number generation out only for performance testing and use the browser random() in production.

Either way, the modulus fix is an awesome find and is a pure win on code complexity, performance and size.

**Comment by MaxBittker** [comment]
no obligation to do any work here, if you want to just land the modulus commit I can take it from there, I do appreciate the PR and the research a lot!

One reason I care about bundle size so much is because the game is so optimized for mobile already, and because I want this to be an exemplar repo for wasm and rust, demonstrating that it's not just for heavy monolithic game engine type stuff

**Comment by MaxBittker** [comment]
I do want that, main concern is squeezing the extra buttons into the layout
and having them be self explanatory .

File: crate/src/lib.rs (hunk 2 above)
Comment (MaxBittker) [reply crate/src/lib.rs]: I do want that, main concern is squeezing the extra buttons into the layout
and having them be self explanatory .

**Comment by MaxBittker** [comment]
gonna merge and then revert the RNG commit, appreciate your help kazimuth!

**Comment by MaxBittker** [comment]
@kazimuth thank you for doing this work 2.5 years ago. It was a mistake to revert it - I was able to go back today and re-incorporate your ideas (PRNG & cache-alignment) and hopefully the game will work better on low spec devices now!

**Pull request merged**